Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean

```diff
@@ -0,0 +1,964 @@
+block of `c`, we may define a function on `Fin n → E` by picking the variables in the `i`-th block
+of `n`, and applying the corresponding coefficient of `p` to these variables. This function is
+called `p.applyOrderedFinpartition c v i` for `v : Fin n → E` and `i : Fin c.k`. -/
+def applyOrderedFinpartition (p : ∀ (i : Fin c.length), E[×c.partSize i]→L[𝕜] F) :
+    (Fin n → E) → Fin c.length → F :=
+  fun v m ↦ p m (v ∘ c.emb m)
+
+lemma applyOrderedFinpartition_apply (p : ∀ (i : Fin c.length), E[×c.partSize i]→L[𝕜] F)
+    (v : Fin n → E) :
+  c.applyOrderedFinpartition p v = (fun m ↦ p m (v ∘ c.emb m)) := rfl
+
+/-- Technical lemma stating how `c.applyOrderedFinpartition` commutes with updating variables. This
+will be the key point to show that functions constructed from `applyOrderedFinpartition` retain
+multilinearity. -/
+theorem applyOrderedFinpartition_update_right
+    (p : ∀ (i : Fin c.length), E[×c.partSize i]→L[𝕜] F)
+    (j : Fin n) (v : Fin n → E) (z : E) :
+    c.applyOrderedFinpartition p (update v j z) =
+      update (c.applyOrderedFinpartition p v) (c.index j)
+        (p (c.index j)
+          (Function.update (v ∘ c.emb (c.index j)) (c.invEmbedding j) z)) := by
+  ext m
+  by_cases h : m = c.index j
+  · rw [h]
+    simp only [applyOrderedFinpartition, update_same]
+    congr
+    rw [← Function.update_comp_eq_of_injective]
+    · simp
+    · exact (c.emb_strictMono (c.index j)).injective
+  · simp only [applyOrderedFinpartition, ne_eq, h, not_false_eq_true,
+      update_noteq]
+    congr
+    apply Function.update_comp_eq_of_not_mem_range
+    have A : Disjoint (range (c.emb m)) (range (c.emb (c.index j))) :=
+      c.disjoint (mem_univ m) (mem_univ (c.index j)) h
+    have : j ∈ range (c.emb (c.index j)) := mem_range.2 ⟨c.invEmbedding j, by simp⟩
+    exact Set.disjoint_right.1 A this
+
+theorem applyOrderedFinpartition_update_left (p : ∀ (i : Fin c.length), E[×c.partSize i]→L[𝕜] F)
+    (m : Fin c.length) (v : Fin n → E) (q : E[×c.partSize m]→L[𝕜] F) :
+    c.applyOrderedFinpartition (update p m q) v
+      = update (c.applyOrderedFinpartition p v) m (q (v ∘ c.emb m)) := by
+  ext d
+  by_cases h : d = m
+  · rw [h]
+    simp [applyOrderedFinpartition]
+  · simp [h, applyOrderedFinpartition]
+
+/-- Given a an ordered finite partition `c` of `n`, a continuous multilinear map `f` in `c.length`
+variable, and for each `m` a continuous multilinear map `p m` in `c.partSize m` variables,
+one can form a continuous multilinear map in `n`
+variables by applying `p m` to each part of the partition, and then
+applying `f` to the resulting vector. It is called `c.compAlongOrderedFinpartition f p`. -/
+def compAlongOrderedFinpartition
+    (f : ContinuousMultilinearMap 𝕜 (fun _i : Fin c.length ↦ F) G)
+    (p : ∀ (i : Fin c.length), E[×c.partSize i]→L[𝕜] F) :
+    E[×n]→L[𝕜] G where
+  toFun v := f (c.applyOrderedFinpartition p v)
+  map_add' v i x y := by
+    cases Subsingleton.elim ‹_› (instDecidableEqFin _)
```

Comment:
Not entirely related to the rest of this PR, but I think this is symptomatic of a very annoying diamond in `DecidableEq (Fin _)` that I've also experienced :(

Reply:
Not really, it's more the definition of multilinear maps, which don't assume `DecidableEq` but instead quantify over all such instances in the fields `map_add'` and `map_smul'`. So, on a concrete type with a concrete `DecidableEq` instance, you always have to make this small dance when proving that something is multilinear.

Reply:
Ah, I see, okay!